crates/cpuid-utils/src/instance_spec.rs

```diff
@@ -0,0 +1,123 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Helpers for converting instance spec CPUID entries into this module's types.
+
+use super::*;
+
+#[cfg(feature = "instance-spec")]
+use propolis_api_types::instance_spec::components::board::CpuidEntry;
+use thiserror::Error;
+
+/// An error that can occur when converting a list of CPUID entries in an
+/// instance spec into a [`CpuidMap`].
+#[derive(Debug, Error)]
+pub enum CpuidMapConversionError {
+    #[error("duplicate leaf and subleaf ({0:x}, {1:?})")]
+    DuplicateLeaf(u32, Option<u32>),
+
+    #[error("leaf {0:x} specified subleaf 0 both explicitly and as None")]
+    SubleafZeroAliased(u32),
+
+    #[error("leaf {0:x} not in standard or extended range")]
+    LeafOutOfRange(u32),
+}
+
+impl From<CpuidMap> for Vec<CpuidEntry> {
+    fn from(value: CpuidMap) -> Self {
+        value
+            .0
+            .into_iter()
+            .map(
+                |(
+                    CpuidIdent { leaf, subleaf },
+                    CpuidValues { eax, ebx, ecx, edx },
+                )| CpuidEntry {
+                    leaf,
+                    subleaf,
+                    eax,
+                    ebx,
+                    ecx,
+                    edx,
+                },
+            )
+            .collect()
+    }
+}
+
+impl TryFrom<Vec<CpuidEntry>> for CpuidMap {
+    type Error = CpuidMapConversionError;
+
+    fn try_from(
+        value: Vec<CpuidEntry>,
+    ) -> Result<Self, CpuidMapConversionError> {
+        let mut map = BTreeMap::new();
+        for CpuidEntry { leaf, subleaf, eax, ebx, ecx, edx } in
+            value.into_iter()
+        {
+            if !(STANDARD_LEAVES.contains(&leaf)
+                || EXTENDED_LEAVES.contains(&leaf))
+            {
+                return Err(CpuidMapConversionError::LeafOutOfRange(leaf));
+            }
+
+            if subleaf.is_none()
+                && map.contains_key(&CpuidIdent::leaf_subleaf(leaf, 0))
+            {
+                return Err(CpuidMapConversionError::SubleafZeroAliased(leaf));
+            }
+
+            if let Some(0) = subleaf {
+                if map.contains_key(&CpuidIdent::leaf(leaf)) {
+                    return Err(CpuidMapConversionError::SubleafZeroAliased(
+                        leaf,
+                    ));
+                }
+            }
```

Comment:
~if provided a `value` like ` vec![CpuidEntry { leaf: N, subleaf: None, .. }, CpuidEntry { leaf: N, subleaf: Some(1), .. }]` the aliasing logic wouldn't notice that and we'd behave in weird (but not obviously disastrous) ways at runtime - the `subleaf: None` would act like a fallback for all `ecx!=1` queries.~

~i don't think any leaves of cpuid need behavior like that, and i think such a situation would be a bug, right?~

https://github.com/oxidecomputer/propolis/pull/782/commits/2ddcd3386730884d1e37e95dc00983ee0c3509eb yay!